python/lib/modeldata/dmod/modeldata/hydrofabric/geopackage_hydrofabric.py

```diff
@@ -0,0 +1,593 @@
+        return cls(layer_names=layer_names,
+                   layer_dataframes=dict([(ln, gpd.read_file(geopackage_file, layer=ln)) for ln in layer_names]),
+                   vpu=vpu,
+                   is_conus=is_conus)
+
+    def __init__(self, layer_names: List[str], layer_dataframes: Dict[str, gpd.GeoDataFrame], vpu: Optional[int] = None,
+                 is_conus: bool = False):
+        self._layer_names: List[str] = layer_names
+        self._dataframes: Dict[str, gpd.GeoDataFrame] = layer_dataframes
+        self._roots = None
+        self._vpu = vpu
+        self._is_conus = is_conus
+
+        #flowpaths = self._dataframes[self._FLOWPATHS_LAYER_NAME]
+        divides = self._dataframes[self._DIVIDES_LAYER_NAME]
+        nexuses = self._dataframes[self._NEXUS_LAYER_NAME]
+
+        col_args = {'col_cat_id': self._DIVIDES_CAT_ID_COL, 'col_nex_id': self._NEXUS_NEX_ID_COL,
+                    'col_to_cat': self._NEXUS_TO_CAT_COL, 'col_to_nex': self._DIVIDES_TO_NEX_COL}
+
+        self._catchments: Dict[str, GeoPackageCatchment] = dict(
+            [(cid, GeoPackageCatchment(cid, self, divides, nexuses, **col_args)) for cid in
+             self.get_all_catchment_ids()])
+
+        self._nexuses: Dict[str, GeoPackageNexus] = dict(
+            [(nid, GeoPackageNexus(nid, self, divides, nexuses, **col_args)) for nid in self.get_all_nexus_ids()])
+
+    def __eq__(self, other):
+        if not isinstance(other, GeoPackageHydrofabric) or self.uid != other.uid:
+            return False
+        if len(self._layer_names) != len(other._layer_names):
+            return False
+        for layer_name in self._layer_names:
+            if layer_name not in other._dataframes:
+                return False
+            elif not self._dataframes[layer_name].equals(other._dataframes[layer_name]):
+                return False
+        return True
+
+    def __hash__(self) -> int:
+        return hash(self.uid)
+
+    def get_all_catchment_ids(self) -> Tuple[str, ...]:
+        """
+        Get ids for all contained catchments.
+
+        Returns
+        -------
+        Tuple[str, ...]
+            Ids for all contained catchments.
+        """
+        return tuple(self._dataframes[self._DIVIDES_LAYER_NAME][self._DIVIDES_CAT_ID_COL].values)
+
+    def get_all_nexus_ids(self) -> Tuple[str, ...]:
+        """
+        Get ids for all contained nexuses.
+
+        Returns
+        -------
+        Tuple[str, ...]
+            Ids for all contained nexuses.
+        """
+        return tuple(self._dataframes[self._NEXUS_LAYER_NAME][self._NEXUS_NEX_ID_COL].values)
+
+    def get_catchment_by_id(self, catchment_id: str) -> Optional[GeoPackageCatchment]:
+        return self._catchments.get(catchment_id)
+
+    def get_nexus_by_id(self, nexus_id: str) -> Optional[GeoPackageNexus]:
+        return self._nexuses.get(nexus_id)
+
+    def get_subset_hydrofabric(self, subset: SubsetDefinition) -> 'GeoPackageHydrofabric':
```

Comment:
> Generally we have thought of changing the details of the hydrofabric as really producing a new hydrofabric.

Right, I agree. We use the term _subset_ and I think of it as a copy, but I think it can be easily taken to mean a view.

> Of course, altering the data (in the original object) isn't strictly forbidden. Perhaps it needs to be. It might get a bit tricky to do properly for `geopandas.GeoDataFrame` objects, though.

We might be able to avoid the behavior completely by using [pandas' copy on write semantics](https://pandas.pydata.org/docs/dev/user_guide/copy_on_write.html#copy-on-write). I was unaware pandas had this feature.